src/graph/InsertVertexExecutor.cpp

```diff
@@ -26,91 +25,132 @@ Status InsertVertexExecutor::prepare() {
             break;
         }
 
-        // TODO(dutor) To support multi-tag insertion
-        auto tagItems = sentence_->tagItems();
-        if (tagItems.size() > 1) {
-            status = Status::Error("Multi-tag not supported yet");
-            break;
-        }
-
-        auto *tagName = tagItems[0]->tagName();
-        properties_ = tagItems[0]->properties();
-
         rows_ = sentence_->rows();
-        // TODO(dutor) To check whether the number of props and values matches.
         if (rows_.empty()) {
             status = Status::Error("VALUES cannot be empty");
             break;
         }
 
+        auto tagItems = sentence_->tagItems();
         overwritable_ = sentence_->overwritable();
+        spaceId_ = ectx()->rctx()->session()->space();
+
+        // Check field name
+        tagIds_.reserve(tagItems.size());
+        schemas_.reserve(tagItems.size());
+        tagProps_.reserve(tagItems.size());
+
+        for (auto& item : tagItems) {
+            auto *tagName = item->tagName();
+            auto tagStatus = ectx()->schemaManager()->toTagID(spaceId_, *tagName);
+            if (!tagStatus.ok()) {
+                return Status::Error("No schema found for `%s'", tagName->c_str());
+            }
 
-        auto spaceId = ectx()->rctx()->session()->space();
-        auto tagStatus = ectx()->schemaManager()->toTagID(spaceId, *tagName);
-        if (!tagStatus.ok()) {
-            status = Status::Error("No schema found for '%s'", tagName->c_str());
-            break;
-        }
-        tagId_ = tagStatus.value();
-        schema_ = ectx()->schemaManager()->getTagSchema(spaceId, tagId_);
-        if (schema_ == nullptr) {
-            status = Status::Error("No schema found for '%s'", tagName->c_str());
-            break;
+            auto tagId = tagStatus.value();
+            auto schema = ectx()->schemaManager()->getTagSchema(spaceId_, tagId);
+            if (schema == nullptr) {
+                return Status::Error("No schema found for `%s'", tagName->c_str());
+            }
+
+            auto props = item->properties();
+            // Now default value is unsupported, props should equal to schema's fields
+            if (schema->getNumFields() != props.size()) {
+                LOG(ERROR) << "props number " << props.size()
+                           << ", schema field number " << schema->getNumFields();
+                return Status::Error("Wrong number of props");
+            }
+
+            tagIds_.emplace_back(tagId);
+            schemas_.emplace_back(schema);
+            tagProps_.emplace_back(props);
+            for (auto fieldIndex = 0u; fieldIndex < schema->getNumFields(); fieldIndex++) {
```

Comment:
this piece of code seems redundant with edge. any idea to improve (common util function?)

Reply:
good advice